src/ExecBackends/MemBackendUmpireImpl.hpp

```diff
@@ -0,0 +1,261 @@
+// product endorsement purposes.
+
+/**
+ * @file MemBackendUmpireImpl.hpp
+ *
+ * @author Cosmin G. Petra <petra1@llnl.gov>, LLNL
+ * 
+ */
+
+/**
+ * This file contains Umpire implementation memory backend.
+ */
+
+#ifndef HIOP_MEM_BCK_UMPIRE
+#define HIOP_MEM_BCK_UMPIRE
+
+#include <ExecSpace.hpp>
+
+#ifdef HIOP_USE_RAJA // can/should be HIOP_USE_UMPIRE
+
+#include <umpire/Allocator.hpp>
+#include <umpire/ResourceManager.hpp>
+
+namespace hiop {
+
+//
+// Memory allocator and deallocator
+//
+template<typename T, typename I>
+struct AllocImpl<MemBackendUmpire, T, I>
+{
+  inline static T* alloc(MemBackendUmpire& mb, const I& n)
+  {
+    auto& resmgr = umpire::ResourceManager::getInstance();
+    umpire::Allocator devalloc  = resmgr.getAllocator(mb.mem_space());
+    return static_cast<T*>(devalloc.allocate(n*sizeof(T)));
+  }
+};
+  
+template<typename T>
+struct DeAllocImpl<MemBackendUmpire, T>
+{
+  inline static void dealloc(MemBackendUmpire& mb, T* p)
+  {
+    auto& resmgr = umpire::ResourceManager::getInstance();
+    umpire::Allocator devalloc  = resmgr.getAllocator(mb.mem_space());
+    devalloc.deallocate(p);
+  }  
+};
+
+//////////////////////////////////////////////////////////////////////////////////////////
+// Transfers
+//////////////////////////////////////////////////////////////////////////////////////////
+
+template<class EXECPOLDEST, class EXECPOLSRC, typename T, typename I>
+struct TransferImpl<MemBackendUmpire, EXECPOLDEST, MemBackendUmpire, EXECPOLSRC, T, I>
+{
+  inline static bool do_it(T* p_dest,
+                           ExecSpace<MemBackendUmpire, EXECPOLDEST>& hwb_dest,
+                           const T* p_src,
+                           const ExecSpace<MemBackendUmpire, EXECPOLSRC>& hwb_src,
+                           const I& n)
+  {
+    auto& rm = umpire::ResourceManager::getInstance();
+
+    if(n>0) {
+      double* src = const_cast<double*>(p_src);
+      rm.copy(p_dest, src, n*sizeof(T));
+    }
+    return true;
+  }
+};
+
+////////////////////////////////////////////////////////////////////////////////////////
+// Transfers to/from Host C++ memory
+////////////////////////////////////////////////////////////////////////////////////////
+template<class EXECPOLDEST, class EXECPOLSRC, typename T, typename I>
+struct TransferImpl<MemBackendCpp, EXECPOLDEST, MemBackendUmpire, EXECPOLSRC, T, I>
+{
+  inline static bool do_it(T* p_dest,
+                           ExecSpace<MemBackendCpp, EXECPOLDEST>& hwb_dest,
+                           const T* p_src,
+                           const ExecSpace<MemBackendUmpire, MemBackendUmpire>& hwb_src,
+                           const I& n)
+  {
+    if(hwb_src.mem_backend().is_host()) {
+      std::memcpy(p_dest, p_src, n*sizeof(T));
+    } else {
+      assert(false && "Transfer BACKENDS(TO:Cpp-host,FROM:umpire) only supported with Umpire mem space host"); 
+      return false;
```

Comment:
I know it would invoke a memory copy, but couldn't we support this with `mem_space=DEVICE` by copying from the device first?

Reply:
we could, but right now I prefer to avoid "under the hood" device/host transfers

Reply:
since we don't have a real need for it rn